### PR TITLE
QEB

### PR DESCRIPTION
**Pull request by damarkian**
Qubit excitation based (QEB) approach for singles, doubles, triples, ... , n-tuples up to arbitrary excitation rank gates in trotterized UCC.

### CODE REVIEW

File: src/qibochem/ansatz/qeb.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        gate_list.append(_g)
+    for _g in mcry_gate:
+        gate_list.append(_g)
+    for _g in reversed(fwd_gates):
+        gate_list.append(_g)
+
+    circuit = Circuit(n_qubits)
+    for _g in gate_list:
+        circuit.add(_g)
+
```
Comment (chmwzc) [inline src/qibochem/ansatz/qeb.py:69]: I refactored the code to replace the for loops with generators, do let me know if there's anything wrong

```suggestion
    i_array = excitation[:n_tuples]
    a_array = excitation[n_tuples:]

    fwd_gates = [gates.CNOT(i_array[-1], _i) for _i in i_array[-2::-1]]
    fwd_gates += [gates.CNOT(a_array[-1], _a) for _a in a_array[-2::-1]]
    fwd_gates.append(gates.CNOT(a_array[-1], i_array[-1]))
    fwd_gates += [gates.X(_ia) for _ia in excitation if _ia not in (i_array[-1], a_array[-1])]

    circuit = Circuit(n_qubits)
    circuit.add(gate for gate in fwd_gates)
    # MCRY
    # multi-controlled RY gate,
    # negative controls i, a
    # positive control on i_n
    mcry_controls = excitation[:-1]
    ry_angle = 2.0 * theta
    circuit.add(gates.RY(a_array[-1], ry_angle).controlled_by(*mcry_controls))

    circuit.add(gate for gate in fwd_gates[::-1])```

File: src/qibochem/ansatz/qeb.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
+import numpy as np
+import qibo
+from qibo import Circuit, gates, models
+
+
+def qeb_circuit(n_qubits, excitation, theta=0.0, trotter_steps=1) -> Circuit:
```
Comment (chmwzc) [inline src/qibochem/ansatz/qeb.py:6]: Does this ansatz involve Trotterization? If not, can take the argument out

```suggestion
def qeb_circuit(n_qubits, excitation, theta=0.0) -> Circuit:
```

File: tests/test_ucc.py [hunk 3]
```diff
@@ -145,6 +146,51 @@ def test_ucc_circuit(excitation, mapping, basis_rotations, coeffs):
...
+        n_terms = len(basis_rotation)
+        pauli_term = SymbolicHamiltonian(
+            symbols.I(n_qubits - 1)
+            * reduce(lambda x, y: x * y, (getattr(symbols, _op)(int(qubit)) for _op, qubit in basis_rotation.split()))
+        )
+        control_circuit += pauli_term.circuit(-coeff * theta)
+    control_result = control_circuit(nshots=1)
+    control_state = control_result.state(True)
+
+    test_circuit = qeb_circuit(n_qubits, excitation, theta=theta, trotter_steps=1)
```
Comment (chmwzc) [inline tests/test_ucc.py:188]: Remove `trotter_steps` argument if not needed

```suggestion
    test_circuit = qeb_circuit(n_qubits, excitation, theta=theta)
```

File: src/qibochem/ansatz/qeb.py (hunk 2 above)
Comment (damarkian) [reply src/qibochem/ansatz/qeb.py:6]: It works as a substitute for the UCC. So yes, it involves trotterization.

File: tests/test_ucc.py (hunk 3 above)
Comment (damarkian) [reply tests/test_ucc.py:188]: As above. https://github.com/qiboteam/qibochem/pull/113#discussion_r1831894579

File: doc/source/tutorials/ansatz.rst [hunk 4]
```diff
@@ -104,7 +104,8 @@ An example of how to build a UCC doubles circuit ansatz for the :math:`H_2` mole
 .. code-block:: python
 
     from qibochem.driver.molecule import Molecule
-    from qibochem.ansatz import hf_circuit, ucc_circuit
+    from qibochem.ansatz.hf_reference import hf_circuit
+    from qibochem.ansatz.ucc import ucc_circuit
```
[the remark this answers is not in the document]
Comment (damarkian) [reply doc/source/tutorials/ansatz.rst]: No reason, and I will go with your suggestion. 

Committed https://github.com/qiboteam/qibochem/pull/113/commits/1958939ae9e6b7b6b44febb86e605a40c85c067a

File: src/qibochem/ansatz/qeb.py (hunk 1 above)
Comment (damarkian) [reply src/qibochem/ansatz/qeb.py]: Thanks. Implemented your suggestion in https://github.com/qiboteam/qibochem/pull/113/commits/677bd382fdbc78e379a39a4636383e8beb55e222

File: src/qibochem/ansatz/qeb.py (hunk 2 above)
Comment (chmwzc) [reply src/qibochem/ansatz/qeb.py:4]: Hmm, but I don't see it being used in the function right now? If it's the same as the original UCC ansatz, then the code probably needs an outside for loop to create a copy of the circuit for each Trotter step, and combine and return everything together.

Comment (damarkian) [reply src/qibochem/ansatz/qeb.py:4]: The QEB gate for any n-tuple is just basically a depth-reduced implementation of the UCC n-tuple gate. It should be directly substituted if one prefers a depth-reduced trotterized UCC. 

My concern is - if one were to use this with trotter-steps > 1, one would have to rewrite code for the outside for loop to create a circuit copy for each Trotter step, combine etc. etc. just to accommodate a simple n-tuple excitation gate substitution. By leaving it in, it allows for QEB to be plugged in directly, no? Maybe using an if-else fork at some point?

**Review by chmwzc — APPROVED** [review]

**Pull request merged**